packages/graphql/src/services/sessions.ts

```diff
@@ -50,6 +183,116 @@ export async function startSession(
   })
 }
 
+interface ActivateSessionBlockArgs {
+  sessionId: string
+  sessionBlockId: number
+}
+
+export async function activateSessionBlock(
+  { sessionId, sessionBlockId }: ActivateSessionBlockArgs,
+  ctx: ContextWithUser
+) {
+  const session = await ctx.prisma.session.findUnique({
+    where: {
+      id: sessionId,
+    },
+    include: { blocks: true },
+  })
+
+  if (!session || session.ownerId !== ctx.user.sub) return null
+
+  const newBlockIndex = session.blocks.findIndex(
+    (block) => block.id === sessionBlockId
+  )
+
+  // if the block is not from the current session, return early
+  if (newBlockIndex === -1) return session
+
+  // if the block is already active, return early
+  if (session.activeBlock === newBlockIndex) return session
+
+  // set the new block to active
+  const updatedSession = await ctx.prisma.session.update({
+    where: {
+      id: sessionId,
+    },
+    data: {
+      activeBlock: newBlockIndex,
+      blocks: {
+        updateMany: [
+          session.activeBlock > -1
+            ? {
+                where: {
+                  id: session.blocks[session.activeBlock].id,
+                },
+                data: {
+                  status: SessionBlockStatus.EXECUTED,
+                },
+              }
+            : undefined,
+          {
+            where: {
+              id: sessionBlockId,
+            },
+            data: {
+              status: SessionBlockStatus.ACTIVE,
+            },
+          },
+        ].filter(Boolean) as any[],
+      },
+    },
+    include: { blocks: true },
+  })
+
+  // get the details and instances of the newly activated session block
+  const newActiveBlock = await ctx.prisma.sessionBlock.findUnique({
+    where: {
+      id: sessionBlockId,
+    },
+    include: {
+      instances: true,
+    },
+  })
+
+  // initialize the cache for the new active block
+  const redisMulti = ctx.redisExec.multi()
+
+  newActiveBlock!.instances.forEach((instance) => {
+    const questionData = instance.questionData!.valueOf() as AllQuestionTypeData
+
+    redisMulti.hmset(`s:${session.id}:i:${instance.id}:info`, {
+      type: questionData.type,
+      namespace: session.namespace,
+    })
+
+    switch (questionData.type) {
+      case QuestionType.SC:
+      case QuestionType.MC: {
+        redisMulti.hmset(`s:${session.id}:i:${instance.id}:results`, {
+          participants: 0,
+          ...(instance.results!.valueOf() as ChoicesQuestionResults).choices,
+        })
+        break
+      }
+
+      case QuestionType.NUMERICAL:
+      case QuestionType.FREE_TEXT: {
+        redisMulti.hmset(`s:${session.id}:i:${instance.id}:results`, {
+          participants: 0,
+        })
+        break
+      }
+    }
+  })
+
+  redisMulti.exec()
+
+  // TODO: if it has been executed already, rehydrate the cache
```

Comment:
This function still needs a lot of work, but should be appropriate for a first "documentation" of the principles :)